### PR TITLE
clh: Pass the tuntap fds to cloud-hypervisor

### PR DESCRIPTION
**Pull request by fidencio**
Instead of passing to cloud-hypervisor the name of the bridge and
letting it open the tuntap device, which is not an allowed operation for
`container_kvm_t` processes, let's just pass the file descriptors to the
VMM, in a similar manner of what's done with QEMU.

Fixes: #3533

And here's an example of the outcome of this short series:
```
[centos@crio runtime]$ kubectl get pods
No resources found in default namespace.

[centos@crio runtime]$ getenforce 
Enforcing

[centos@crio runtime]$ kubectl apply -f ~/pod.yaml 
pod/nginx created

[centos@crio runtime]$ kubectl get pods
NAME    READY   STATUS    RESTARTS   AGE
nginx   1/1     Running   0          8s

[centos@crio runtime]$ ps auxZ | grep cloud-hypervisor
system_u:system_r:container_kvm_t:s0:c185,c571 root 167962 3.2  0.6 2256596 99336 ? Sl 20:21   0:00 /usr/bin/cloud-hypervisor --api-socket /run/vc/vm/e03001e1d25806805c61c9facb5a1ec90e83c62c5f327ef08e8c9305d72fcdfa/clh-api.sock
```

### CODE REVIEW

**Review by likebreath — APPROVED** [review]

**Comment by likebreath** [comment]
@fidencio `ch-remote` is your friend, for example, you can check the VM state with:

```
ch-remote --api-socket /path/to/clh-api.sock info
```

File: src/runtime/virtcontainers/clh.go
Comment (amshinde) [inline src/runtime/virtcontainers/clh.go:31]: Change looks good to me @fidencio. Just a nitpick about the commit message here. Replace
/Instead of passing to cloud-hypervisor the name of the bridge/Instead of passing to cloud-hypervisor the name of the tuntap device/

Since we were initially passing the name of the tuntap device before this change.

File: src/runtime/virtcontainers/clh.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/clh.go:1166]: Is there a CH issue on this topic that this comment could reference, so we don't forget about this?

File: src/runtime/virtcontainers/clh.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/clh.go:1185]: Could you add a test to `virtcontainers/clh_test.go` for this?

File: src/runtime/virtcontainers/network_test.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/network_test.go:199]: How about adding:

```diff
diff --git a/src/runtime/virtcontainers/network_test.go b/src/runtime/virtcontainers/network_test.go
index d98d04169..fef60f571 100644
--- a/src/runtime/virtcontainers/network_test.go
+++ b/src/runtime/virtcontainers/network_test.go
@@ -199,12 +199,20 @@ func TestCreateGetTunTapLink(t *testing.T) {
        assert.NotNil(tapLink)
        assert.NotZero(len(fds))

+       tuntap, ok := tapLink.(*netlink.Tuntap)
+       assert.True(ok)
+       assert.Equal(tuntap.Queues, 1)
+
        tapName = "testtap1"
        tapLink, fds, err = createLink(netHandle, tapName, &netlink.Tuntap{}, 1)
        assert.NoError(err)
        assert.NotNil(tapLink)
        assert.NotZero(len(fds))

+       tuntap, ok = tapLink.(*netlink.Tuntap)
+       assert.True(ok)
+       assert.Equal(tuntap.Queues, 1)
+
        tapLink, err = getLinkByName(netHandle, tapName, &netlink.Tuntap{})
        assert.NoError(err)
```

File: src/runtime/virtcontainers/clh.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/clh.go:1167]: It would be good to add an explicit test for this: set the `FD_CLOEXEC` flag and check to ensure it gets removed.

**Review by jodh-intel — COMMENTED** [review]
Thanks @fidencio - a few comments...

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

**Comment by fidencio** [comment]
Folks, thanks for the reviews but this is still a draft and I'm working on changes on the cloud-hypervisor side that will be needed here.  I will revisit this, and ping y'all once this PR matures enough to not be a draft anymore. :-)

**Comment by jodh-intel** [comment]
Any update on this @fidencio?

**Comment by jodh-intel** [comment]
@fidencio - could you update to resolve the conflicts and address the comments?

**Comment by jodh-intel** [comment]
Any update on this @fidencio?

**Comment by arronwy** [comment]
Can one of the admins verify this patch?

**Comment by fidencio** [comment]
@jodh, this became a lower priority due to TDX / Cloud Hypervisor related work.
I will rework this one in the coming weeks.

**Pull request closed without merging**
